runtime/serde/serde-xml/common/src/aws/smithy/kotlin/runtime/serde/xml/XmlDeserializer.kt

```diff
@@ -98,10 +104,19 @@ internal class XmlMapDeserializer(
     private val mapTrait = descriptor.findTrait<XmlMapName>() ?: XmlMapName.Default
 
     override fun hasNextEntry(): Boolean {
-        // Seek to either the entry or key token depending on the flatness of the map
+        val compareTo = when (descriptor.hasTrait<Flattened>()) {
+            // Prefer seeking to XmlSerialName if the trait exists
+            true -> when (descriptor.hasTrait<XmlSerialName>()) {
+                true -> descriptor.expectTrait<XmlSerialName>().name
+                false -> mapTrait.key
+            }
+            false -> mapTrait.entry
+        }
```

Comment:
**Style**: I generally find `when`/`true`/`false` to be less readable than `if`/`else` or null-coalescing. In this case, `when (descriptor.hasTrait<XmlSerialName>()) { ... }` can be replaced with:

```kotlin
descriptor.findTrait<XmlSerialName>()?.name ?: mapTrait.key
```